src/pybind/mgr/rook/module.py

```diff
@@ -492,10 +492,9 @@ def create_osds(self, drive_group):
         if drive_group.data_directories:
             targets += drive_group.data_directories
 
-        def execute(all_hosts_):
-            # type: (List[orchestrator.HostSpec]) -> orchestrator.Completion
-            all_hosts = [h.hostname for h in all_hosts_]
-            matching_hosts = drive_group.placement.pattern_matches_hosts(all_hosts)
+        def execute():
+            # type: () -> orchestrator.Completion
+            matching_hosts = drive_group.placement.filter_matching_hosts(self.get_hosts)
```

Comment:
For this, you need to split `self.get_hosts` into to methods:

```python
class RookOrchestrator:
    def get_hosts(self) -> orchestrator.Completion:
        ...

    def _get_hosts(self) -> List[HostSpec]
        ...
```

then you can do

```suggestion
            matching_hosts = drive_group.placement.filter_matching_hosts(self._get_hosts)
```

Reply:
Isn't the method I'm using already returning `List[HostSpec]` -> https://github.com/ceph/ceph/blob/9ea2ec7472bb8a580cf1bc4823298364716d26b5/src/pybind/mgr/rook/module.py#L253-L256 ?

Reply:
Ah, `deferred_read` ..  the type annotation is just wrong here. Fixing

Reply:
note that mypy and pycharm use different mechanics for type annotations:

consider
```
def to_int(f: Callable[[Any], str]) -> int:
    def wrapper(arg: Any) -> int:
        return f(arg)
    return wrapper

@to_int
def works_in_mypy(arg: Any) -> str:
    return str(arg)

@to_int
def works_in_pycharm(arg: Any) -> int:
    return str(arg)
```

I'm using the mypy style till now in the code base.